ENSNode dockerfile (#186)
namehash/ensnode: apps/ensnode/README.md

  - removes centralized dependency on the CCIP Gateway
  - flaky test experience with .cb.id name gateway
  - also helps indexer discovery

## Dockerfile

> Instructions for building the Docker image for ENSNode.

To build the Docker image, navigate to the top of the monorepo and run the following command:

```bash

> docker build -f apps/ensnode/Dockerfile -t ensnode .

Reviewer: i think the easier approach is to add the `image` field in the docker compose and have people run `docker compose build` from the root of the monorepo. they can also specify a container like `docker compose build ensnode`

Author: I would believe that we are talking about two separate ideas here (Docker and Docker compose). Here we are just showing how to set up a single docker container as this is the context of this README (it's only describing ENSNode not a whole system).
Docker compose description at the top of the monorepo is a guide on how to set up "all" components to run on a single machine in a more holistic way. I can think about some devs or clients who are not willing to do that and would only pay attention to the ENSNode component

```

This command will use the Dockerfile located in the `apps/ensnode` directory to create the image.

### Important Note
The Dockerfile expects the build context to be at the top of the monorepo. This means that all the necessary files and directories it needs to access during the build process should be available from that location. If you run the build command from a different directory, the Dockerfile may not find the required files, leading to errors. Keeping the context at the top ensures everything is organized and accessible for a smooth build process.

> ### Team and Docker Image Repository

Reviewer 2: ```suggestion
### Public Docker Image Repository
```

Is this change fair?

Author: Yep, all good. Adjusted

> The team responsible for this repository is publishing its own Docker image under the following Docker image repository: [ENSNode image repository](https://github.com/namehash/ensnode/pkgs/container/ensnode%2Fensnode). This allows users to easily pull the pre-built image without needing to build it themselves.

Reviewer 2: ```suggestion
NameHash Labs publishes prebuilt ENSNode Docker images for the community under a [Docker image repository](https://github.com/namehash/ensnode/pkgs/container/ensnode%2Fensnode). You can use this to easily pull a prebuilt image without needing to build it yourself.
```

A few questions:
1. Are the above suggestions fair?
2. If someone from the community wanted to use one of these prebuilt images, how would they do it? Sorry if this is a stupid question.

> ### Running the Application

Reviewer 2: Sorry, but a few stupid questions:

1. I've noted the instructions for how to customize the env vars in the overall `docker-compose.yml` file. But what if I'm just building the Docker container exclusively for ENSNode using the Dockerfile in this `apps/ensnode` directory? How do I customize my env vars here? Do I do it when building the container and/or when I run the container?
2. How does it work if someone downloads the prebuilt Docker image from the image repository on GitHub? How do they customize the env vars there?

Author: 1. I've added mention about setting up env variables for container run. They are defined during contianer creation, no need to provide them during image build.
2. The same way. You are providing them when runnning `docker run` see an example provided.

> To run the application using Docker, you can use the following command:

Reviewer 2: ```suggestion
After building the Docker image, you can run the application with the following command:
```

Is that change fair? The main goal in this proposed change is to make sure its very explicit that you need to finish building the image first before this command will work.

Author: Adjusted

```bash

> docker run -p 42069:42069 ensnode

Reviewer: likewise, i think it would be easiest to standardize on docker-compose, to avoid having extra config flags floating around. i think the suggestion should be `docker compose up ensnode`

Reviewer: and to that end, i'd actually prefer to remove the `docker:build` and `docker:push` scripts in the individual package folders and perhaps instead add a script at the root package.json like

```
"docker:build": "docker compose build"
"docker:push": "docker compose push"
```

and in the ci we can use it like `pnpm run docker:build ensnode && pnpm run docker:push ensnode` to build/push a specific image

Author: Look at my comment above about Docker/Docker-compose. 
About `docker:build` and `docker:push`. I would also dispose of those scripts.
I'm sceptical however about relying on `docker compose build` to actually build the image before pushing. That would mean we want to build all images at once which might not be desired.
I would completely remove any Docker or docker-compose related scripts from `package.json` and rely on scripts provided in CI workflow (use CLI for Docker related tasks).

```

> This command will start the container and map port `42069` from the container to your local machine, allowing you to access the application.

Reviewer 2: ```suggestion
This command will start the container for ENSNode and map port `42069` from the container to your local machine, allowing you to access ENSNode.
```

Is that fair? The goal in this suggestion is to be more explicit on exactly what application this starts, since there's more than one application.

Author: Yep, adjusted

### Docker Compose

> For a more convenient setup, you can also use Docker Compose. Refer to the Docker Compose usage instructions at the top of the monorepo for details on how to run the application with Docker Compose, which simplifies the process of managing multi-container applications.

Reviewer 2: A few ideas here:

1. Suggest changing "Refer to the Docker Compose usage instructions at the top of the monorepo" to an link to the associated README.md file that is an anchor link to the specific section you're referring to here.
2. As I understand, if someone just builds and runs the ENSNode docker image by itself then we should be telling them that they need to be sure to manually configure the env vars (please see my questions above related to how we communicate this instruction) for dependent services: Postgres and ENSRainbow. I'm worried about people running ENSNode through any strategy before they manually configure env vars. For each of the journeys we document for people to get it running, please ensure we always explicitly encourage configuring env vars before they actually run it.

Author: 1. Added links to correct files.
2. Added instructions for setting up env vars and creating their own DB instance for indexer
